### PR TITLE
Add `PShow`

### PR DESCRIPTION
**Pull request by srid**
Things to do,

- [x] Derive instances for other types
- [x] Add `ptraceShowId` and the like
- [x] `PByteString` instance (as hexstring)
- [x] `PShow`: quote the output via bytestring using utf8 encoding builtins

### CODE REVIEW

**Comment by TotallyNotChase** [comment]
if we're adding a `Show` like typeclass - it should probably only exist in development mode (due to the size and cost). Or am I misunderstanding its practical usecases?

**Comment by emiflake** [comment]
> if we're adding a `Show` like typeclass - it should probably only exist in development mode

I think if we can eliminate all traces anyway, the use of `Show` in a trace would be eliminated anyway, right?

**Comment by srid** [comment]
`pshow` is mainly useful in `ptrace*` - so yes, in non-development mode all of this should vanish (unless you are using `pshow` _outside_ of tracing for some reason).

I'd expect users to just use `ptraceShowId` and the like, in conjunction with `deriving anyclass PShow` on their data types, without using `pshow` directly.

**Review by L-as — CHANGES_REQUESTED** [review]
nice

File: Plutarch/Trace.hs [hunk 1]
```diff
@@ -34,6 +44,14 @@ ptrace s a = pforce $ ptrace' # s # pdelay a
 ptrace _ a = a
 #endif
 
+-- | Like Haskell's `traceShowId` but for Plutarch
+ptraceShowId :: PShow a => Term s a -> Term s a
+#ifdef Development
+ptraceShowId x = ptrace (pshow x) x
```
Comment (TotallyNotChase) [inline Plutarch/Trace.hs:50]: maybe this function should be called `ptraceShow`

Comment (srid) [reply Plutarch/Trace.hs:50]: Well, it is analogous to Haskell's `traceShowId` - I kept the naming for consistency.

File: Plutarch/Show.hs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+-- | Return the string representation of a Plutarch value
+pshow :: PShow a => Term s a -> Term s PString
+pshow = pshow' False
+
+instance PShow PString where
+  -- FIXME: We ideally want to wrap the string in quotes, and escape it
+  -- appropriately just like Haskell's Show instance. However, there appears to
+  -- be no way to do this using Plutus' builtins. So we will just return the
+  -- string as is.
+  pshow' _ = id
```
[the remark this answers is not in the document]
Comment (srid) [reply Plutarch/Show.hs:70]: Yes, I already use `<>` liberally - but `pshow' _ x = "\"" <> x "\""` would not be correct as it wouldn't escape the double quotes inside of `x`.

File: Plutarch/Show.hs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+-- | Generic version of `pshow`
+gpshow ::
+  forall a s.
+  (PGeneric s a, HasDatatypeInfo (a s), PlutusType a, All2 PShow (PCode s a)) =>
+  Bool ->
+  Term s (a :--> PString)
+gpshow wrap =
+  phoistAcyclic $
+    punsafeAsClosedTerm @s $
```
[the remark this answers is not in the document]
Comment (srid) [reply Plutarch/Show.hs:115]: @L-as See also this PR which does the same: https://github.com/Plutonomicon/plutarch/pull/335/files#diff-bdc1d03d75f8e4a1a716fbda46bf5b0fcf50f7a1ede7d7730430ce0e8bcf15b7R151-R154

File: Plutarch/Show.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Show.hs:70]: oh wow that's a neat idea

File: Plutarch/Show.hs [hunk 4]
```diff
@@ -99,10 +101,40 @@ instance PShow PInteger where
           pcase perror digit $
             flip fmap [0 .. 9] $ \(x :: Integer) ->
               (pconstant x, pconstant (T.pack . show $ x))
-      pcase :: PEq a => Term s b -> Term s a -> [(Term s a, Term s b)] -> Term s b
-      pcase otherwise x = \case
-        [] -> otherwise
-        ((x', r) : cs) -> pif (x #== x') r $ pcase otherwise x cs
+
+instance PShow PByteString where
+  pshow' _ x = "0x" <> showByteString # x
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Show.hs:124]: the `<>`

File: Plutarch/ByteString.hs [hunk 5]
```diff
@@ -74,6 +78,23 @@ pbyteStr = pconstant
 pconsBS :: Term s (PInteger :--> PByteString :--> PByteString)
 pconsBS = punsafeBuiltin PLC.ConsByteString
 
+-- | Case matching on bytestring, as if a list.
+pelimBS ::
```
Comment (L-as) [inline Plutarch/ByteString.hs:82]: I don't think we should export this function. It's not directly necessary for Plutarch, nor is it even a good abstraction to use.

File: Plutarch/Show.hs [hunk 6]
```diff
@@ -65,11 +65,29 @@ pshow :: PShow a => Term s a -> Term s PString
 pshow = pshow' False
 
 instance PShow PString where
-  -- FIXME: We ideally want to wrap the string in quotes, and escape it
-  -- appropriately just like Haskell's Show instance. However, there appears to
-  -- be no way to do this using Plutus' builtins. So we will just return the
-  -- string as is.
-  pshow' _ = id
+  pshow' _ x = "\"" <> pshowStr # x <> "\""
```
Comment (L-as) [inline Plutarch/Show.hs:68]: You can hoist this

**Review by L-as — CHANGES_REQUESTED** [review]
Looks mostly good to me

**Comment by srid** [comment]
Hoisted PString, PByteString impl, as well as unexposed pelimBS.

File: Plutarch/Unit.hs [hunk 7]
```diff
@@ -39,4 +38,4 @@ instance Monoid (Term s PUnit) where
   mempty = pcon PUnit
 
 instance PShow PUnit where
-  pshow' _ x = pmatch x $ \PUnit -> "()"
+  pshow' _ _ = "()"
```
Comment (L-as) [inline Plutarch/Unit.hs:41]: this is actually an oversight in the other instances, and I don't think this is actually a "fix", see #341

**Review by L-as — APPROVED** [review]
LGTM, we should fix the PUnit stuff in another PR

**Comment by L-as** [comment]
@srid please resolve conflicts

**Review by L-as — APPROVED** [review]
LGTM

**Pull request merged**